src/bokeh/models/callbacks.py

```diff
@@ -101,21 +113,79 @@ class CustomJS(Callback):
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
 
-    args = Dict(String, AnyRef, help="""
+    args = Dict(String, AnyRef)(default={}, help="""
     A mapping of names to Python objects. In particular those can be bokeh's models.
     These objects are made available to the callback's code snippet as the values of
     named parameters to the callback.
     """)
 
-    code = String(default="", help="""
-    A snippet of JavaScript code to execute in the browser. The
-    code is made into the body of a function, and all of of the named objects in
+    code = Required(String)(help="""
+    A snippet of JavaScript code to execute in the browser.
+
+    This can be interpreted either as a JavaScript function or a module, depending
+    on the ``module`` property:
+
+    1. A JS function.
+
+    The code is made into the body of a function, and all of of the named objects in
     ``args`` are available as parameters that the code can use. Additionally,
     a ``cb_obj`` parameter contains the object that triggered the callback
     and an optional ``cb_data`` parameter that contains any tool-specific data
     (i.e. mouse coordinates and hovered glyph indices for the ``HoverTool``).
+
+    2. An ES module.
+
+    A JavaScript module (ESM) exporting a default function with the following
+    signature:
+
+    .. code-block: javascript
+
+        export default function(args, obj, data) {
+            // program logic
+        }
+
+    where ``args`` is a key-value mapping of user-provided parameters, ``obj``
+    refers to the object that triggered the callback, and ``data`` is a key-value
+    mapping of optional parameters provided by the caller.
+
+    This function can be an asynchronous function (``async function``).
+    """)
+
+    module = Either(Auto, Bool)(default="auto", help="""
```

Comment:
Aren't these usually
```
    module = Either(Auto, Bool, default="auto", help="""
```
I don't have a problem with this per-se, but I'd like to be consisten.

Reply:
I reverted this one for the sake of consistency. However, going forward, I think we should switch to writing complex property types like this, because this makes it easier to delineate the type more easily when reading the code, and when modifying the type.